### PR TITLE
fix(test): close all open connections in abstract_client test

### PR DESCRIPTION
**Pull request by MaximoLiberata**
Continue the work made in #1911

### CODE REVIEW

**Comment by robertsLando** [comment]
@MaximoLiberata Let me know when this is ready for review

**Comment by MaximoLiberata** [comment]
@robertsLando How can I configure my computer to run this test https://github.com/mqttjs/MQTT.js/actions/runs/10217524082/job/28271502928?pr=1917? It's a new error for me.

I don't know if this change 8d154b0e81e848f9fa295946bdfd5d379c0af7ae directly affects the new error or maybe the previous PR

**Comment by robertsLando** [comment]
I have a feel that's happening because I removed mqtt from deps in one of the last commits I pushed to master, it's just a tought but I'm not on my PC right now so cannot say much more. 

In order to run it just check how workflow is setup, if on windows you may need to use WSL

**Comment by MaximoLiberata** [comment]
I could to run the test. The test might throw an error because `hivemq` rejects the connection and the client never connect, sometimes the client can connect even if `hivemq` rejected the connection. So, depend of `hivemq` if the test fails sometimes.

**Comment by robertsLando** [comment]
Maybe we could switch to a self hosted broker like I did for browser tests with aedes cli?

**Comment by MaximoLiberata** [comment]
Sincerely I don't know. I could try to move it to a self hosted broker later.

**Comment by MaximoLiberata** [comment]
I couldn't to reproduce the [hanging bug](https://github.com/mqttjs/MQTT.js/actions/runs/10182861315/job/28166435859) of the keepalive test in my computer, but I could to reproduce it only in the GitHub Actions.


It seems the test fails because the `client` never publish the packages even it's connected. Using this [commit](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/commit/807d5e5cd1114d00700a6f68cae551a2f85bdc67) to show the logs of the test' states I could to get two cases:

[Case 1 - GitHub Action](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10231223067/job/28306931838)
[Case 2 - GitHub Action](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10231218526/job/28306921431)

<details>
<summary>Case 1 - Test Pass</summary>

```
   ▶ keepalive
    ✔ should send ping at keepalive interval (10.629001ms)
    ✔ should not shift ping on publish (3.967672ms)
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'connack',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    sessionPresent: false,
    reasonCode: 0
  }
}
CONNECT { reschedulePings: true, callCount: 0, received: 0 }
PUBLISH {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 46591
  }
}
PUBLISH {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 46592
  }
}
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 46591,
    reasonCode: 0
  }
}
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 1,
  received: 1,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 46592,
    reasonCode: 0
  }
}
    ✔ should reschedule pings if publishing at a higher rate than keepalive and reschedulePings===true (8.595212ms)
PACKETRECEIVE {
  reschedulePings: false,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'connack',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    sessionPresent: false,
    reasonCode: 0
  }
}
CONNECT { reschedulePings: false, callCount: 0, received: 0 }
PUBLISH {
  reschedulePings: false,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 23409
  }
}
PUBLISH {
  reschedulePings: false,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 23410
  }
}
PACKETRECEIVE {
  reschedulePings: false,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 23409,
    reasonCode: 0
  }
}
PACKETRECEIVE {
  reschedulePings: false,
  callCount: 0,
  received: 1,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 23410,
    reasonCode: 0
  }
}
    ✔ should not reschedule pings if publishing at a higher rate than keepalive and reschedulePings===false (4.602472ms)
    ✔ should shift ping on pingresp when reschedulePings===true (3.483994ms)
PACKETRECEIVE {
  reschedulePings: false,
  callCount: 0,
  received: 2,
  packet: Packet {
    cmd: 'connack',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    sessionPresent: false,
    reasonCode: 0
  }
}
    ✔ should shift ping on pingresp when reschedulePings===false (3.745445ms)
  ▶ keepalive (35.51644ms)
```

</details>


<details>
<summary>Case 2 - Test Fails</summary>

```
  ▶ keepalive
    ✔ should send ping at keepalive interval (10.889386ms)
    ✔ should not shift ping on publish (4.545398ms)
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'connack',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    sessionPresent: false,
    reasonCode: 0
  }
}
CONNECT { reschedulePings: true, callCount: 0, received: 0 }
PUBLISH {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 62056
  }
}
PUBLISH {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'publish',
    retain: false,
    qos: 1,
    dup: false,
    length: 11,
    topic: 'foo',
    payload: <Buffer 62 61 72>,
    messageId: 62057
  }
}
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 62056,
    reasonCode: 0
  }
}
PACKETRECEIVE {
  reschedulePings: true,
  callCount: 1,
  received: 1,
  packet: Packet {
    cmd: 'puback',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    messageId: 62057,
    reasonCode: 0
  }
}
    ✔ should reschedule pings if publishing at a higher rate than keepalive and reschedulePings===true (9.137764ms)
PACKETRECEIVE {
  reschedulePings: false,
  callCount: 0,
  received: 0,
  packet: Packet {
    cmd: 'connack',
    retain: false,
    qos: 0,
    dup: false,
    length: 3,
    topic: null,
    payload: null,
    sessionPresent: false,
    reasonCode: 0
  }
}
CONNECT { reschedulePings: false, callCount: 0, received: 0 }
✖ /home/runner/work/MQTT-ReactNative-Support/MQTT-ReactNative-Support/test/client_mqtt5.ts (60001.160737ms)
  'test timed out after 60000ms'
```

</details>


@robertsLando  Do you have any idea of the bug's reason?

**Comment by robertsLando** [comment]
> @robertsLando Do you have any idea why client do not publish the packages?

Usually in such situations the reason is a race condition, using fake timers there has been a pain, I had to put clock.tick [here](https://github.com/mqttjs/MQTT.js/blob/b39ee126634810cbb301ea36dc4e39de5023ed56/test/abstract_client.ts#L2143) and [there](https://github.com/mqttjs/MQTT.js/blob/b39ee126634810cbb301ea36dc4e39de5023ed56/test/abstract_client.ts#L2150) in order to prevent this kind of issues, my feel is that sometimes the stack trace (in slower machines like GH actions) could be a bit different for some reason and the clock ticks are not enough in order to trigger the next tick from the call queue so the event is not emitted. A possible (dirty) solution could be to call a clock.tick inside an interval in order to make the process go on when this happens

**Comment by robertsLando** [comment]
> @robertsLando Do you have any idea why client do not publish the packages?

Usually in such situations the reason is a race condition, using fake timers there has been a pain, I had to put clock.tick [here](https://github.com/mqttjs/MQTT.js/blob/b39ee126634810cbb301ea36dc4e39de5023ed56/test/abstract_client.ts#L2143) and [there](https://github.com/mqttjs/MQTT.js/blob/b39ee126634810cbb301ea36dc4e39de5023ed56/test/abstract_client.ts#L2150) in order to prevent this kind of issues, my feel is that sometimes the stack trace (in slower machines like GH actions) could be a bit different for some reason and the clock ticks are not enough in order to trigger the next tick from the call queue so the event is not emitted. A possible (dirty) solution could be to call a clock.tick inside an interval in order to make the process go on when this happens

**Comment by MaximoLiberata** [comment]
I don't know if this will throw an idea to resolve the bug, but these are the mqtt logs (using `console.log` in `log` option) that GH Actions threw:

[GH Action - Test Pass](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10241402343)
[GH Action - Test Fails](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10241397313)

And here is the difference between two mqtt logs when Pass and when Fails https://www.diffchecker.com/902zCyLn/ (The left side is when Pass and the right side is when fails)

Also, `client` throw an error when keepalive is reached and only happens when `reschedulePings===false`

```
Error: Keepalive timeout
    at onKeepaliveTimeout (/home/max/Downloads/MQTT/src/lib/client.ts:2143:22)
    at <anonymous> (/home/max/Downloads/MQTT/src/lib/KeepaliveManager.ts:108:17)
    at callTimer (/home/max/Downloads/MQTT/node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:783:24)
    at doTickInner (/home/max/Downloads/MQTT/node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:1380:29)
    at doTick (/home/max/Downloads/MQTT/node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:1461:20)
    at Object.tick (/home/max/Downloads/MQTT/node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:1469:20)
    at <anonymous> (/home/max/Downloads/MQTT/test/abstract_client.ts:2150:15)
```

**Comment by robertsLando** [comment]
FYI you can check this box when running tests in order to see debug logs: 



Anyway I'm quite sure the issue is related to the thing I told you here: https://github.com/mqttjs/MQTT.js/pull/1917#issuecomment-2268458739

**Comment by MaximoLiberata** [comment]
@robertsLando I think I could to prevent or fix the hanging bug in `reschedulePings` test. I ran 80 GH actions to check it and now I only receive other new bugs. Maybe they're correlational to this change or not, but I hope that this change will be an advance.

[GitHub Actions](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/workflows/mqttjs-test.yml)
Start 481 and end 560

[GH Action bug 1](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10252008630/job/28361384797)
[GH Action bug 2](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10252053238/job/28361538533)
[GH Action bug 3](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/runs/10252130434/job/28361814845)

**Comment by MaximoLiberata** [comment]
@robertsLando Could you check the changes, please?

I think we got it, I ran 100 GH Actions and not one of them failed. I removed or changed some `setTimeout`, they probably were affected by `clock` because parallel process. 

[GH Actions](https://github.com/MaximoLiberata/MQTT-ReactNative-Support/actions/workflows/mqttjs-test.yml)
Start 620 and end 720

File: test/abstract_client.ts [hunk 1]
```diff
@@ -2101,64 +2126,79 @@ export default function abstractTest(server, config, ports) {
...
+								locaClock.restore()
+							}
+						},
+					)
 
-				client.on('packetreceive', (packet) => {
-					if (packet.cmd === 'puback') {
-						process.nextTick(() => {
-							clock.tick(intervalMs)
+					const locaClock = sinon.useFakeTimers({
```
Comment (robertsLando) [inline test/abstract_client.ts:2151]: Just a small typo here, fix it then we can merge :) 

```suggestion
					const localClock = sinon.useFakeTimers({
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: test/abstract_client.ts (hunk 1 above)
Comment (MaximoLiberata) [reply test/abstract_client.ts]: Done.

**Review by robertsLando — COMMENTED** [review]
LGTM 🚀

**Pull request merged**